src/transformStateWithClones.js

```diff
@@ -8,6 +8,32 @@
  */
 function transformStateWithClones(state, actions) {
   // write code here
+  let previousState = { ...state };
+  const history = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(previousState, action.extraData);
```

Comment:
Instead of mutating 'previousState' directly inside the 'addProperties' case, you should create a new state copy with the added properties to ensure immutability.